Automattic/jetpack
# Autoloader: Preemptively load unknown plugins from cache

Fixes https://github.com/woocommerce/woocommerce/issues/27608

#### Changes proposed in this Pull Request:

One of the limitations of the autoloader is that it is only able to detect plugins using the `active_plugins` option, `active_sitewide_plugins` option, and request parameters. If a plugin is an `mu-plugin` or loaded via code, it will be added to the `$jetpack_autoloader_activating_plugin` global and trigger a reset. This comes with a performance penalty as well as the potential for odd behavior when two classes from different versions of the same package are autoloaded.

This PR introduces a cache file in the `wp-content/cache` folder containing a list of plugins that are known to the autoloader on shutdown. By saving this cache on shutdown we're able to preemptively handle autoloaders in plugins that we aren't aware of initially. 

This solution is not perfect however, as it causes the autoloader to consider plugins that may not necessarily be active anymore. We aim for eventual consistency instead by only writing plugins that we've discovered to the cache, so the request after a plugin is deactivated, it will be removed from the cache. My thinking is that the presence of a deactivated plugin's class files shouldn't cause any problems. One odd thing that might happen is that a newer autoloader is used in a deactivated plugin, however, the cache invalidation will mean this only happens a single time.

I've also taken the opportunity to finish the refactor I started in #16819. One of the core motivations behind this refactoring was to decrease the cognitive cost of working with the autoloader. Although the code may be simple it hid a lot of complexity and required deep thought into how different elements might interact. Here are some highlights:
* High-level integration tests that check common cross-plugin conflict scenarios
* `CacheHandler` for reading/writing filesystem cache files. This can enable other optimizations like saving the reconciled classmap.
* `Container` for managing all of the dependencies in the package.
* `Plugin_Locator` to centralize all of the plugin lookup logic and generalize certain shared actions
* `Autoloader` to serve as a new entry point
* `Hook_Manager` shared by all autoloaders so that resets can remove hooks added by other autoloaders

#### Jetpack product discussion
N/A

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:

* Make sure Jetpack works by itself
* Make sure Jetpack works alongside WooCommerce
* Make sure Jetpack as an `mu-plugin` works alongside WooCommerce

There's arguably a large matrix of possible ways to test the autoloader, which I attempted to cover most if not all of in the `Test_Autoloader_Scenarios` integration tests. If you see any tests that we should add, please note!

#### Proposed changelog entry for your changes:

* Autoloader: Added support for preemptively loading unknown plugins from a cache file.
